yash-builtin/src/command/identify.rs

```diff
@@ -0,0 +1,582 @@
+
+//! Command identifying semantics
+
+use super::search::SearchEnv;
+use super::Identify;
+use crate::command::Category;
+use crate::common::{output, report_failure, to_single_message};
+use std::borrow::Cow;
+use std::ffi::CStr;
+use std::ffi::CString;
+use std::ffi::OsStr;
+use std::os::unix::ffi::OsStrExt as _;
+use std::os::unix::ffi::OsStringExt as _;
+use std::path::PathBuf;
+use yash_env::builtin::Type;
+use yash_env::semantics::ExitStatus;
+use yash_env::semantics::Field;
+use yash_env::Env;
+use yash_env::System;
+use yash_quote::quoted;
+use yash_semantics::command_search::search;
+use yash_semantics::command_search::Target;
+use yash_syntax::parser::lex::Keyword;
+use yash_syntax::source::pretty::Annotation;
+use yash_syntax::source::pretty::AnnotationType;
+use yash_syntax::source::pretty::MessageBase;
+
+/// Error object for the command not found
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct NotFound<'a> {
+    /// Command name that was not found
+    pub name: &'a Field,
+}
+
+impl MessageBase for NotFound<'_> {
+    fn message_title(&self) -> Cow<str> {
+        "command not found".into()
+    }
+
+    fn main_annotation(&self) -> Annotation<'_> {
+        let label = format!("{}: not found", self.name.value).into();
+        let location = &self.name.origin;
+        Annotation::new(AnnotationType::Error, label, location)
+    }
+}
+
+/// Environment for [normalizing a target](normalize_target).
+trait NormalizeEnv {
+    fn is_executable_file(&self, path: &CStr) -> bool;
+    fn pwd(&self) -> Result<PathBuf, ()>;
+}
+
+impl NormalizeEnv for Env {
+    #[inline]
+    fn is_executable_file(&self, path: &CStr) -> bool {
+        self.system.is_executable_file(path)
+    }
+
+    fn pwd(&self) -> Result<PathBuf, ()> {
+        match self.get_pwd_if_correct() {
+            Some(pwd) => Ok(pwd.into()),
+            None => self.system.getcwd().map_err(|_| ()),
+        }
+    }
+}
+
+/// Updates the target to make it suitable for the [description](describe_target).
+///
+/// This function makes sure any path contained in the target is absolute and
+/// names an executable file. If the path cannot be normalized, this function
+/// returns an error.
+fn normalize_target<E: NormalizeEnv>(env: &E, target: &mut Target) -> Result<(), ()> {
+    match target {
+        Target::Function(_) | Target::Builtin { path: None, .. } => Ok(()),
+
+        Target::External { path }
+        | Target::Builtin {
+            path: Some(path), ..
+        } => {
+            if !env.is_executable_file(path) {
+                return Err(());
+            }
+            if !path.as_bytes().starts_with(b"/") {
+                let mut absolute_path = env.pwd()?;
+                absolute_path.push(OsStr::from_bytes(path.as_bytes()));
+                *path = CString::new(absolute_path.into_os_string().into_vec()).map_err(|_| ())?;
+            }
+            Ok(())
+        }
+    }
+}
```

Comment:
Thanks for the suggestion, but I chose to keep the current function signature because this function is internal in this module and the returned error value does not need to convey any information. For clarification, I extended the doc comment in commit 52df77efc4e745b41f251c458b02bea3cfb7da89.